### PR TITLE
prov/rxm: allow core providers to decide if flow control should be enabled

### PR DESCRIPTION
**Pull request by ooststep**
This set of changes is a proposal to address #5907 and enables core providers to decide if credit based flow control should be enabled.  

With this implementation, verbs will disable flow control if shared receive queue is used.

### CODE REVIEW

File: include/ofi_util.h [hunk 1]
```diff
@@ -1008,6 +1008,7 @@ struct ofi_ops_flow_ctrl {
 	size_t	size;
 	void	(*set_threshold)(struct fid_domain *domain, uint64_t threshold);
 	void	(*add_credits)(struct fid_ep *ep, uint64_t credits);
+	ssize_t	(*enable)(struct fid_ep *ep);
```
Comment (shefty) [inline include/ofi_util.h:1011]: This should return int.  ssize_t is mostly used with data transfer operations (for possible future extensibility).  int is used for other cases.

File: prov/rxm/src/rxm_conn.c [hunk 2]
```diff
@@ -897,14 +903,9 @@ static int rxm_msg_ep_open(struct rxm_ep *rxm_ep, struct fi_info *msg_info,
 	}
 
 	ret = fi_open_ops(&rxm_domain->msg_domain->fid, OFI_OPS_FLOW_CTRL, 0,
-			  (void **) &flow_ctrl_ops, &msg_ep->fid);
+			  (void **) &flow_ctrl_ops, NULL);
```
Comment (shefty) [inline prov/rxm/src/rxm_conn.c:906]: Should this call move now?  Maybe to where the msg domain is opened?

File: prov/rxm/src/rxm_conn.c [hunk 3]
```diff
@@ -942,6 +943,15 @@ static int rxm_msg_ep_open(struct rxm_ep *rxm_ep, struct fi_info *msg_info,
 		goto err;
 	}
 
+	ret = rxm_ep->flow_ctrl_ops->enable(msg_ep);
+	if (!ret) {
+		rxm_ep->flow_ctrl_ops->set_threshold(
+			rxm_domain->msg_domain,
+			rxm_ep->msg_info->rx_attr->size / 2);
+		rxm_ep->flow_ctrl_ops->set_send_handler(rxm_domain->msg_domain,
+							rxm_send_credits);
```
Comment (shefty) [inline prov/rxm/src/rxm_conn.c:952]: The set_handler and set_send_handler should move next to the open_ops call.

File: prov/verbs/src/verbs_domain.c [hunk 4]
```diff
@@ -58,25 +58,35 @@ static void vrb_set_credit_handler(struct fid_domain *domain_fid,
 	domain->send_credits = credit_handler;
 }
 
+static ssize_t vrb_enable_ep_flow_ctrl(struct fid_ep *ep_fid)
+{
+	struct vrb_ep *ep = container_of(ep_fid, struct vrb_ep, util_ep.ep_fid);
+	// only enable if we are not using SRQ
+	if (ep->srq_ep) {
```
Comment (shefty) [inline prov/verbs/src/verbs_domain.c:65]: Doesn't this mean that the endpoint is using SRQ?

File: include/ofi_util.h [hunk 5]
```diff
@@ -205,6 +205,7 @@ struct util_domain {
 	struct ofi_mr_map	mr_map;
 	enum fi_threading	threading;
 	enum fi_progress	data_progress;
+	struct ofi_ops_flow_ctrl *flow_ctrl_ops;
```
Comment (shefty) [inline include/ofi_util.h:208]: Why was this moved from rxm to the general code?

File: prov/verbs/src/verbs_ep.c [hunk 6]
```diff
@@ -331,6 +331,7 @@ vrb_alloc_init_ep(struct fi_info *info, struct vrb_domain *domain,
 		goto err3;
 	}
 
+	ep->threshold = UINT64_MAX; /* disables RQ flow control */
```
Comment (shefty) [inline prov/verbs/src/verbs_ep.c:334]: Please move this to the same location peer_rq_credits is initialized.

File: include/ofi_util.h (hunk 5 above)
Comment (ooststep) [reply include/ofi_util.h:208]: oops. yeah i'll move that back to into rxm. this was an oversight on my part.

File: prov/verbs/src/verbs_domain.c [hunk 7]
```diff
@@ -58,25 +55,35 @@ static void vrb_set_credit_handler(struct fid_domain *domain_fid,
 	domain->send_credits = credit_handler;
 }
 
+static int vrb_enable_ep_flow_ctrl(struct fid_ep *ep_fid)
+{
+	struct vrb_ep *ep = container_of(ep_fid, struct vrb_ep, util_ep.ep_fid);
+	// only enable if we are not using SRQ
+	if (!ep->srq_ep) {
```
Comment (shefty) [inline prov/verbs/src/verbs_domain.c:62]: Should also check that this is a msg ep (rc qp).

File: prov/verbs/src/verbs_domain.c [hunk 8]
```diff
@@ -58,25 +55,35 @@ static void vrb_set_credit_handler(struct fid_domain *domain_fid,
 	domain->send_credits = credit_handler;
 }
 
+static int vrb_enable_ep_flow_ctrl(struct fid_ep *ep_fid)
+{
+	struct vrb_ep *ep = container_of(ep_fid, struct vrb_ep, util_ep.ep_fid);
+	// only enable if we are not using SRQ
+	if (!ep->srq_ep && ep->ibv_qp->qp_type == IBV_QPT_RC) {
```
Comment (shefty) [inline prov/verbs/src/verbs_domain.c:62]: nit: I'd go a step further and verify that ep->ibv_qp is set prior to dereferencing it.

**Comment by shefty** [comment]
thanks!
